File: examples/matrix_mul/matrix_mul.c
/*
 * C matrix-multiply benchmark - counterpart to matrix_mul.mettle
 *
 * Build: build.bat (or: gcc -O3 -o matrix_mul_c.exe matrix_mul.c -lkernel32)
 */

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>

#include "../bench_time.h"

#define N 32
#define N2 (N * N)

static volatile int64_t g_bench_sink;

static int mat_idx(int row, int col) {
    return row * N + col;
}

static void fill_matrix(int32_t *m, int seed) {
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            m[mat_idx(row, col)] = (int32_t)((row * 131 + col * 17 + seed) % 251);
        }
    }
}

static void matmul(const int32_t *a, const int32_t *b, int32_t *c) {
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            int32_t sum = 0;
            for (int k = 0; k < N; k++) {
                sum += a[mat_idx(row, k)] * b[mat_idx(k, col)];
            }
            c[mat_idx(row, col)] = sum;
        }
    }
}

static int64_t trace_matrix(const int32_t *m) {
    int64_t sum = 0;
    for (int i = 0; i < N; i++) {
        sum += (int64_t)m[mat_idx(i, i)];
    }
    return sum;
}

int main(void) {
    const int passes = 200;
    int32_t a[N2];
    int32_t b[N2];
    int32_t c[N2];

    fill_matrix(a, 3);
    fill_matrix(b, 7);

    printf("Matrix multiply: 32x32 int32 (naive)\n");

    matmul(a, b, c);
    int64_t check = trace_matrix(c);
    printf("Trace = %" PRId64 "\n", check);
    printf("Benchmark: %d passes (matmul)\n", passes);

    uint64_t t0 = bench_time_us();
    int64_t bench_sum = 0;
    for (int p = 0; p < passes; p++) {
        matmul(a, b, c);
        bench_sum += trace_matrix(c);
    }
    uint64_t elapsed_us = bench_time_us() - t0;

    g_bench_sink ^= bench_sum;

    printf("Bench sum = %" PRId64 "\n", bench_sum);
    printf("Time: %" PRIu64 " us\n", elapsed_us);
    printf("Per pass: ~%" PRIu64 " us\n", elapsed_us / (uint64_t)passes);

    return 0;
}

File: examples/matrix_mul/matrix_mul.rs
// Rust matrix-multiply benchmark - same workload as matrix_mul.c.

#[path = "../bench_time.rs"]
mod bench_time;

use bench_time::bench_time_us;
use std::hint::black_box;

const N: usize = 32;
const N2: usize = N * N;
const PASSES: i32 = 200;

fn mat_idx(row: usize, col: usize) -> usize {
    row * N + col
}

fn fill_matrix(m: &mut [i32; N2], seed: i32) {
    for row in 0..N {
        for col in 0..N {
            m[mat_idx(row, col)] = (row as i32 * 131 + col as i32 * 17 + seed).rem_euclid(251);
        }
    }
}

fn matmul(a: &[i32; N2], b: &[i32; N2], c: &mut [i32; N2]) {
    for row in 0..N {
        for col in 0..N {
            let mut sum = 0i32;
            for k in 0..N {
                sum += a[mat_idx(row, k)] * b[mat_idx(k, col)];
            }
            c[mat_idx(row, col)] = sum;
        }
    }
}

fn trace_matrix(m: &[i32; N2]) -> i64 {
    let mut sum = 0i64;
    for i in 0..N {
        sum += i64::from(m[mat_idx(i, i)]);
    }
    sum
}

fn main() {
    let mut a = [0i32; N2];
    let mut b = [0i32; N2];
    let mut c = [0i32; N2];

    fill_matrix(&mut a, 3);
    fill_matrix(&mut b, 7);

    println!("Matrix multiply: 32x32 int32 (naive)");

    matmul(&a, &b, &mut c);
    let check = trace_matrix(&c);
    println!("Trace = {check}");
    println!("Benchmark: {PASSES} passes (matmul)");

    let t0 = bench_time_us();
    let mut bench_sum = 0i64;
    for p in 0..PASSES {
        matmul(&a, &b, &mut c);
        bench_sum += trace_matrix(&c);
        black_box(p);
    }
    let elapsed_us = bench_time_us() - t0;

    println!("Bench sum = {bench_sum}");
    println!("Time: {elapsed_us} us");
    println!("Per pass: ~{} us", elapsed_us / PASSES as u64);
}
